phoenicis-tools/src/main/java/org/phoenicis/tools/system/GraphicsPropertiesFetcher.java

```diff
@@ -0,0 +1,116 @@
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+package org.phoenicis.tools.system;
+
+import org.phoenicis.configuration.security.Safe;
+import org.phoenicis.tools.system.GraphicsProperties;
+
+import org.lwjgl.*;
+import org.lwjgl.glfw.*;
+import org.lwjgl.opengl.*;
+import org.lwjgl.vulkan.*;
+
+import static org.lwjgl.glfw.Callbacks.*;
+import static org.lwjgl.glfw.GLFW.*;
+import static org.lwjgl.opengl.GL11.*;
+import static org.lwjgl.system.MemoryUtil.*;
+import static org.lwjgl.glfw.GLFWVulkan.*;
+
+// Code from LWJGL tutorial
+@Safe
+public class GraphicsPropertiesFetcher {
+    private long window = NULL;
+
+    /**
+     * Create an invisible glfx window and context, from which info will be retrieved
+     */
+    private void init() {
+        GLFWErrorCallback.createPrint(System.err).set();
+
+        if (!glfwInit())
+            throw new IllegalStateException("Unable to initialize GLFW");
+
+        glfwDefaultWindowHints();
+        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
+
+        this.window = glfwCreateWindow(300, 300, "Test Window", NULL, NULL);
+        if (this.window == NULL)
+            throw new RuntimeException("Failed to create the GLFW window for testing graphic card capabilities");
+    }
+
+    /**
+     * Destroy glfw window and context
+     */
+    private void terminate() {
+        glfwFreeCallbacks(this.window);
+        glfwDestroyWindow(this.window);
+
+        glfwTerminate();
+        glfwSetErrorCallback(null).free();
+        this.window = NULL;
+    }
+
+    /**
+     * Fetch graphics card vendor and OpenGL version
+     */
+    private void FetchVendorRendererOpenGLVersion(GraphicsProperties graphicsProperties) {
+        glfwMakeContextCurrent(this.window);
+
+        GL.createCapabilities();
+
+        graphicsProperties.vendor = glGetString(GL_VENDOR);
+        graphicsProperties.renderer = glGetString(GL_RENDERER);
+        graphicsProperties.openGLVersion = glGetString(GL_VERSION);
+        graphicsProperties.openGLVersion = graphicsProperties.openGLVersion.substring(0,
+                graphicsProperties.openGLVersion.indexOf(' '));
+    }
+
+    /**
+     * Fetch Vulkan version
+     */
+    private void FetchVulkanVersion(GraphicsProperties graphicsProperties) {
+        if (!glfwVulkanSupported()) {
+            return;
+        }
+
+        int version = VK.getInstanceVersionSupported();
+
+        graphicsProperties.vulkanVersion = String.valueOf(version >> 22) + "." +
+                String.valueOf((version >> 12) & 0x3ff) + "." +
+                String.valueOf(version & 0xfff);
+    }
+
+    /**
+     * Fetch the current graphics properties (vendor, OpenGL version, Vulkan version) of the system
+     * 
+     * @return The current graphics properties inside a GraphicsProperties object
+     */
+    public GraphicsProperties GetProperties() {
+        GraphicsProperties graphicsProperties = new GraphicsProperties();
```

Comment:
`final`

Reply:
Is not `final` similar to `const` in C++ ?

Reply:
Not exactly. In C++ `const` can be used to enforce two things:
- to enforce that the `const` value itself can't be replaced, i.e. the following is forbidden:

   ```c++
   const X *x = ...;

   x = new X(...); // forbidden
   ```
- the `const` value can't be changed, i.e. the following is forbidden:
   
   ```c++
   const X *x = ...;
 
   x->a = ...; // forbidden
   ```

In Java `final` can be used to enforce that the `final` value itself can't be replaced.